be/src/exec/olap_scan_prepare.cpp

```diff
@@ -171,23 +179,192 @@ static bool get_predicate_value(ObjectPool* obj_pool, const SlotDescriptor& slot
     return true;
 }
 
-template <LogicalType SlotType, typename RangeValueType>
-void OlapScanConjunctsManager::normalize_in_or_equal_predicate(const SlotDescriptor& slot,
-                                                               ColumnValueRange<RangeValueType>* range) {
+static std::vector<ExprContextContainer> build_expr_context_containers(const std::vector<ExprContext*>& expr_contexts) {
+    std::vector<ExprContextContainer> containers;
+    for (auto* expr_ctx : expr_contexts) {
+        containers.emplace_back(expr_ctx);
+    }
+    return containers;
+}
+
+static std::vector<RawExprContainer> build_raw_expr_containers(const std::vector<Expr*>& exprs) {
+    std::vector<RawExprContainer> containers;
+    for (auto* expr : exprs) {
+        containers.emplace_back(expr);
+    }
+    return containers;
+}
+
+template <bool Inverted>
+static TExprOpcode::type maybe_invert_in_and_equal_op(const TExprOpcode::type op) {
+    if constexpr (!Inverted) {
+        return op;
+    } else {
+        switch (op) {
+        case TExprOpcode::FILTER_IN:
+            return TExprOpcode::FILTER_NOT_IN;
+        case TExprOpcode::FILTER_NOT_IN:
+            return TExprOpcode::FILTER_IN;
+        case TExprOpcode::EQ:
+            return TExprOpcode::NE;
+        case TExprOpcode::NE:
+            return TExprOpcode::EQ;
+
+        default:
+            return op;
+        }
+    }
+}
+
+// ------------------------------------------------------------------------------------
+// ChunkPredicateBuilder
+// ------------------------------------------------------------------------------------
+
+RawExprContainer::RawExprContainer(Expr* root_expr) : root_expr(root_expr) {}
+Expr* RawExprContainer::root() const {
+    return get_root_expr(root_expr);
+}
+StatusOr<ExprContext*> RawExprContainer::expr_context(ObjectPool* obj_pool, RuntimeState* state) const {
+    if (new_expr_ctx == nullptr) {
```

Comment:
so if we use a RawExprContainer,  then its expr is cloned and new ExprContext is constructed?

it seems Expr need not be cloned?  so cloning a new one, then put self-managed obj_pool to avoid accident deallocation?

stupid suggest, can we use another name, for example,  
ExprPlus,
ExprContextPlus,

or BoxedExpr
or BoxedExprContext;

or define Boxed template?
```
template <typename E>
struct Boxed<E> {
ExprContext* expr_context(...);
}
```

Reply:
1. Why we need `ExprContextContainer` and `RawExprContainer`?

Because the root `ChunkPredicateBuilder` accepts `vector<ExprContext>` directly from `scan_node.conjuncts`, 

but the non-root `ChunkPredicateBuilder` accepts `vector<Expr>` from a `{And,Or}CompoundExpr::children`. We need create `ExprContext` for each `Expr` of `{And,Or}CompoundExpr::children`, if and only if all the children could be parsed by `ChunkPredicateBuilder`.

2. Why `Expr` need not be cloned?

Because we will create a new `ExprContext` for each child `Expr` of `{And,Or}CompoundExpr`, so there are two `ExprContext` owns this child `Expr`:
- The `OlapScanOperatorFactory::_conjunct_exprs` owns the original `ExprContext` of `{And,Or}CompoundExpr`.
- The `OlapChunkSource::_predicates` owns the new `ExprContext` of each child `Expr`  of `{And,Or}CompoundExpr`.
This causes that each child `Expr` will be released two times.

Reply:
I see, so we always keep the invariant two expr never owns an identical expr context, am I right?  frankly speak, a best design should be one-copy-expr and multi-copy-exprcontext, expr is just compuation logic, and expr context keeps some private state.  however, we handle this not correctly, so it causes twice-release issue.

Reply:
Yes, we always keep the invariant two **expr contexts** never owns an identical **expr**, because we cannot handle this correctly.

Maybe we could optimize the `expr context` and `expr` logic in the future.